module/zfs/dmu_object.c

```diff
@@ -80,6 +80,17 @@ dmu_object_alloc_impl(objset_t *os, dmu_object_type_t ot, int blocksize,
 	if (dnodes_per_chunk > L1_dnode_count)
 		dnodes_per_chunk = L1_dnode_count;
 
+	/*
+	 * The caller requested the dnode be returned as a performance
+	 * optimization in order to avoid releasing the hold only to
+	 * immediately reacquire it.  Since they caller is responsible
+	 * for releasing the hold they must provide the tag.
+	 */
+	if (allocated_dnode != NULL) {
+		ASSERT3P(tag, !=, NULL);
+		tag = FTAG;
```

Comment:
Should this be in an `else`?  (also, do we test with reference_tracking_enabled, which I would think would fail?  IIRC we do ztest with it enabled on illumos)

Reply:
Good catch, yes it should be.  Fixed.

Regarding `reference_tracking_enable` the same is true for Linux.  

```c
#ifdef _KERNEL
int reference_tracking_enable = FALSE; /* runs out of memory too easily */
#else
int reference_tracking_enable = TRUE;
#endif
```